MarkupsToModel/MRML/vtkMRMLMarkupsToModelNode.cxx

```diff
@@ -267,165 +248,318 @@ void vtkMRMLMarkupsToModelNode::SetAndObserveMarkupsNodeID( const char* markupsI
+  }
+  this->SetAndObserveNodeReferenceID( OUTPUT_MODEL_ROLE, modelId );
+}
+
 void vtkMRMLMarkupsToModelNode::ProcessMRMLEvents( vtkObject *caller, unsigned long event, void *callData )
 {
   vtkMRMLNode* callerNode = vtkMRMLNode::SafeDownCast( caller );
   if ( callerNode == NULL ) return;
 
-  if (this->GetMarkupsNode() && this->GetMarkupsNode()==caller)
+  if ( this->GetMarkupsNode() && this->GetMarkupsNode()==caller )
   {
     this->InvokeCustomModifiedEvent(InputDataModifiedEvent);
   }
 }
 
-void vtkMRMLMarkupsToModelNode::RemoveAllMarkups()
+std::string vtkMRMLMarkupsToModelNode::GetModelNodeName()
 {
-  vtkMRMLMarkupsFiducialNode* markupsNode = vtkMRMLMarkupsFiducialNode::SafeDownCast( this->GetNodeReference( MARKUPS_ROLE ) );
-  markupsNode->RemoveAllMarkups();
+  vtkMRMLModelNode* modelNode = vtkMRMLModelNode::SafeDownCast( this->GetNodeReference( OUTPUT_MODEL_ROLE ) );
+  if ( modelNode == NULL )
+  {
+    return std::string( this->GetID() ).append( "Model" );
+  }
+  else
+  {
+    return std::string( modelNode->GetName() );
+  }
 }
 
-void vtkMRMLMarkupsToModelNode::RemoveLastMarkup()
+std::string vtkMRMLMarkupsToModelNode::GetModelDisplayNodeName()
 {
-  vtkMRMLMarkupsFiducialNode* markupsNode = vtkMRMLMarkupsFiducialNode::SafeDownCast( this->GetNodeReference( MARKUPS_ROLE ) );
-  if(markupsNode->GetNumberOfFiducials()>0)
+  vtkMRMLModelNode* modelNode = vtkMRMLModelNode::SafeDownCast( this->GetNodeReference( OUTPUT_MODEL_ROLE ) );
+  if ( modelNode == NULL )
+  {
+    return std::string( this->GetID() ).append( "ModelDisplay" );
+  }
+  
+  vtkMRMLModelDisplayNode* displayNode = vtkMRMLModelDisplayNode::SafeDownCast( modelNode->GetDisplayNode() );
+  if ( displayNode == NULL )
   {
-    markupsNode->RemoveMarkup(markupsNode->GetNumberOfFiducials()-1);
+    return std::string( this->GetID() ).append( "ModelDisplay" );
   }
+
+  return std::string( displayNode->GetName() );
 }
 
-std::string vtkMRMLMarkupsToModelNode::GetModelNodeName()
+std::string vtkMRMLMarkupsToModelNode::GetMarkupsDisplayNodeName()
 {
-  if(this->ModelNode == NULL)
+  vtkMRMLModelNode* markupsNode = vtkMRMLModelNode::SafeDownCast( this->GetNodeReference( OUTPUT_MODEL_ROLE ) );
+  if ( markupsNode == NULL )
   {
-    return std::string(this->GetID()).append("Model");
+    return std::string( this->GetID() ).append( "MarkupsDisplay" );
   }
-  else
+  
+  vtkMRMLMarkupsDisplayNode* displayNode = vtkMRMLMarkupsDisplayNode::SafeDownCast( markupsNode->GetDisplayNode() );
+  if ( displayNode == NULL )
   {
-    return std::string(this->ModelNode->GetName());
+    return std::string( this->GetID() ).append( "MarkupsDisplay" );
   }
+
+  return std::string( displayNode->GetName() );
 }
 
-std::string vtkMRMLMarkupsToModelNode::GetDisplayNodeName()
+void vtkMRMLMarkupsToModelNode::SetOutputOpacity( double outputOpacity )
 {
-  if(this->ModelNode == NULL)
+  vtkMRMLModelNode* modelNode = vtkMRMLModelNode::SafeDownCast( this->GetNodeReference( OUTPUT_MODEL_ROLE ) );
```

Comment:
There is lots of copy-paste code for getting/creating the display node. It should get a dedicated method (but not in this class).